src/qibo/core/terms.py

```diff
@@ -0,0 +1,273 @@
+import sympy
+from qibo import gates, K
+from qibo.config import raise_error, EINSUM_CHARS
+
+
+class HamiltonianTerm:
+    """Single term in a :class:`qibo.core.hamiltonians.SymbolicHamiltonian`.
+
+    This Hamiltonian is represented by a sum of such terms which are held in
+    its ``.terms`` attribute which is a list.
+
+    Args:
+        matrix (np.ndarray): Full matrix corresponding to the term representation
+            in the computational basis. Has size (2^n, 2^n) where n is the
+            number of target qubits of this term.
+        q (int): Target qubit ids.
+    """
+
+    def __init__(self, matrix, *q):
+        self.target_qubits = tuple(q)
+        self._gate = None
+        self.hamiltonian = None
+        if not (matrix is None or isinstance(matrix, K.qnp.numeric_types) or
+                isinstance(matrix, K.qnp.tensor_types)):
+            raise_error(TypeError, "Invalid type {} of symbol matrix."
+                                   "".format(type(matrix)))
+        self._matrix = matrix
+
+    @property
+    def matrix(self):
+        return self._matrix
+
+    @property
+    def gate(self):
+        """Qibo gate that implements the action of the term on states."""
+        if self._gate is None:
+            self._gate = gates.Unitary(self.matrix, *self.target_qubits)
+        return self._gate
+
+    def exp(self, dt):
```

Comment:
Maybe add some check for dt > 0.

Reply:
We could add that but in principle it is okay if one calls this method with dt < 0 as all it does is calculating exp(-i dt H) which is a valid unitary for all dt. Perhaps the name dt is not the best here, perhaps we could switch it to an arbitrary x?

Reply:
Ok, I agree. Replacing with `x` is fine.